File: src/types/preview.ts
/**
 * Types TypeScript pour le service de génération de previews
 * Phase 2.3 - Génération de Previews (Pyramide d'Images)
 */

/**
 * Types de previews disponibles (pyramide d'images)
 */
export enum PreviewType {
  /** Thumbnail pour grille (240px bord long, JPEG q75) */
  Thumbnail = 'thumbnail',
  /** Standard pour affichage plein écran (1440px bord long, JPEG q85) */
  Standard = 'standard',
  /** 1:1 pour zoom pixel (résolution native, JPEG q90) */
  OneToOne = 'one_to_one',
}

/**
 * Configuration pour la génération de previews
 */
export interface PreviewConfig {
  /** Répertoire du catalogue (contient Previews.lrdata/) */
  catalog_dir: string;
  /** Nombre de threads pour traitement parallèle */
  parallel_threads: number;
  /** Timeout pour génération (secondes) */
  generation_timeout: number;
  /** Utiliser libvips si disponible, sinon image crate */
  use_libvips: boolean;
}

/**
 * Résultat de la génération d'une preview
 */
export interface PreviewResult {
  /** Chemin vers le fichier preview généré */
  path: string;
  /** Type de preview */
  preview_type: PreviewType;
  /** Dimensions réelles (largeur, hauteur) */
  size: [number, number];
  /** Taille du fichier en octets */
  file_size: number;
  /** Temps de génération en millisecondes */
  generation_time: number;
  /** Hash BLAKE3 du fichier original */
  source_hash: string;
  /** Timestamp de génération */
  generated_at: string;
}

/**
 * Métadonnées d'une preview dans la base de données previews.db
 */
export interface PreviewRecord {
  /** ID unique de la preview */
  id: number;
  /** Hash BLAKE3 du fichier source */
  source_hash: string;
  /** Type de preview */
  preview_type: PreviewType;
  /** Chemin relatif depuis Previews.lrdata/ */
  relative_path: string;
  /** Dimensions réelles (largeur, hauteur) */
  width: number;
  height: number;
  /** Taille du fichier en octets */
  file_size: number;
  /** Qualité JPEG utilisée */
  jpeg_quality: number;
  /** Date de génération */
  generated_at: string;
  /** Date de dernier accès */
  last_accessed: string;
  /** Nombre d'accès */
  access_count: number;
}

/**
 * Nouvelle preview à insérer dans la base
 */
export interface NewPreviewRecord {
  source_hash: string;
  preview_type: PreviewType;
  relative_path: string;
  width: number;
  height: number;
  file_size: number;
  jpeg_quality: number;
}

/**
 * Informations sur le cache de previews
 */
export interface PreviewCacheInfo {
  /** Nombre total de previews en cache */
  total_previews: number;
  /** Taille totale du cache (octets) */
  total_size: number;
  /** Nombre de thumbnails */
  thumbnail_count: number;
  /** Nombre de previews standard */
  preview_count: number;
  /** Dernier cleanup */
  last_cleanup: string | null;
}

/**
 * Statistiques de génération batch
 */
export interface BatchPreviewStats {
  /** ID unique du batch */
  batch_id: string;
  /** Nombre total de fichiers traités */
  total_files: number;
  /** Nombre de succès */
  successful_count: number;
  /** Nombre d'échecs */
  failed_count: number;
  /** Nombre de fichiers skip (déjà en cache) */
  skipped_count: number;
  /** Temps total de traitement en millisecondes */
  total_duration: number;
  /** Temps moyen par fichier en millisecondes */
  avg_time_per_file: number;
  /** Timestamp de début */
  started_at: string;
  /** Timestamp de fin */
  completed_at: string | null;
}

/**
 * Erreurs spécifiques au service preview
 */
export type PreviewError =
  | { type: 'unsupported_format'; format: string }
  | { type: 'corrupted_file'; path: string }
  | { type: 'processing_error'; message: string }
  | { type: 'write_error'; path: string }
  | { type: 'cache_error'; message: string }
  | { type: 'generation_timeout'; timeout: number }
  | { type: 'out_of_memory' }
  | { type: 'io_error'; message: string };

/**
 * État d'une preview dans le cache
 */
export enum PreviewStatus {
  /** Preview générée et valide */
  Ready = 'ready',
  /** En cours de génération */
  Generating = 'generating',
  /** Erreur de génération */
  Failed = 'failed',
  /** Obsolète (fichier source modifié) */
  Stale = 'stale',
}

/**
 * Métadonnées d'une preview en cache
 */
export interface PreviewMetadata {
  /** Hash BLAKE3 du fichier source */
  source_hash: string;
  /** Type de preview */
  preview_type: PreviewType;
  /** Chemin du fichier preview */
  path: string;
  /** État actuel */
  status: PreviewStatus;
  /** Taille du fichier preview */
  file_size: number;
  /** Dimensions */
  size: [number, number];
  /** Date de génération */
  generated_at: string;
  /** Dernier accès */
  last_accessed: string;
  /** Nombre d'accès */
  access_count: number;
}

/**
 * Événement de progression de génération
 */
export interface PreviewProgressEvent {
  /** Type de l'événement */
  type: 'preview_type_started' | 'preview_type_completed' | 'batch_completed';
  /** Type de preview (pour les événements de type) */
  preview_type?: PreviewType;
  /** Index actuel (pour preview_type_started) */
  current?: number;
  /** Total (pour preview_type_started) */
  total?: number;
  /** Nombre de succès (pour preview_type_completed) */
  successful?: number;
  /** Nombre d'échecs (pour preview_type_completed) */
  failed?: number;
  /** Nombre de fichiers skip (pour preview_type_completed) */
  skipped?: number;
  /** Statistiques complètes (pour batch_completed) */
  stats?: BatchPreviewStats;
}

/**
 * Configuration du cache cleanup
 */
export interface CacheCleanupConfig {
  /** Taille maximale du cache (octets) */
  max_cache_size: number;
  /** Âge maximal des previews (jours) */
  max_age_days: number;
  /** Nombre maximal de previews par type */
  max_previews_per_type: number;
  /** Fréquence de cleanup (heures) */
  cleanup_interval_hours: number;
}

/**
 * Options pour la génération de preview pyramidale
 */
export interface PreviewPyramidOptions {
  /** Générer tous les types de previews */
  generate_all: boolean;
  /** Types spécifiques à générer */
  preview_types?: PreviewType[];
  /** Forcer la régénération même si existant */
  force_regenerate: boolean;
  /** Émettre des événements de progression */
  emit_progress: boolean;
}

/**
 * Résultat de la génération pyramidale
 */
export interface PreviewPyramidResult {
  /** Hash du fichier source */
  source_hash: string;
  /** Résultats par type de preview */
  results: PreviewResult[];
  /** Temps total de génération */
  total_generation_time: number;
  /** Timestamp de génération */
  generated_at: string;
}

/**
 * Statistiques de performance du service preview
 */
export interface PreviewPerformanceStats {
  /** Temps moyen par type de preview */
  avg_generation_time: {
    thumbnail: number;
    standard: number;
    one_to_one: number;
  };
  /** Débit de traitement (previews/seconde) */
  throughput: number;
  /** Taux de succès */
  success_rate: number;
  /** Utilisation mémoire (octets) */
  memory_usage: number;
  /** Nombre de threads actifs */
  active_threads: number;
}

/**
 * Options de recherche de previews
 */
export interface PreviewSearchOptions {
  /** Types de previews à inclure */
  preview_types?: PreviewType[];
  /** Filtrer par date de génération */
  generated_after?: string;
  generated_before?: string;
  /** Filtrer par taille de fichier */
  min_file_size?: number;
  max_file_size?: number;
  /** Ordre de tri */
  sort_by?: 'generated_at' | 'file_size' | 'access_count';
  /** Ordre de tri */
  sort_order?: 'asc' | 'desc';
  /** Pagination */
  offset?: number;
  /** Nombre maximum de résultats */
  limit?: number;
}

/**
 * Résultat de recherche de previews
 */
export interface PreviewSearchResult {
  /** Previews trouvées */
  previews: PreviewRecord[];
  /** Nombre total de résultats */
  total_count: number;
  /** Offset utilisé */
  offset: number;
  /** Nombre maximum demandé */
  limit: number;
}

/**
 * État du service preview
 */
export interface PreviewServiceStatus {
  /** Service initialisé */
  initialized: boolean;
  /** Configuration actuelle */
  config: PreviewConfig;
  /** Informations sur le cache */
  cache_info: PreviewCacheInfo;
  /** Statistiques de performance */
  performance_stats: PreviewPerformanceStats;
  /** Nombre de générations en cours */
  active_generations: number;
  /** Dernière erreur */
  last_error: PreviewError | null;
}

/**
 * Événements émis par le service preview
 */
export type PreviewServiceEvent =
  | { type: 'preview_generated'; result: PreviewResult }
  | { type: 'preview_error'; error: PreviewError; source_hash: string }
  | { type: 'batch_started'; batch_id: string; total_files: number }
  | { type: 'batch_completed'; stats: BatchPreviewStats }
  | { type: 'cache_cleanup'; removed_count: number; freed_space: number }
  | { type: 'service_status'; status: PreviewServiceStatus };

/**
 * Types d'export pour les previews
 */
export enum PreviewExportFormat {
  JPEG = 'jpeg',
  PNG = 'png',
  TIFF = 'tiff',
  WEBP = 'webp',
}

/**
 * Options d'export de preview
 */
export interface PreviewExportOptions {
  /** Format d'export */
  format: PreviewExportFormat;
  /** Qualité (pour JPEG/WebP) */
  quality?: number;
  /** Redimensionner */
  resize?: {
    width: number;
    height: number;
  };
  /** Appliquer des filtres */
  filters?: {
    brightness?: number;
    contrast?: number;
    saturation?: number;
  };
}

/**
 * Résultat d'export de preview
 */
export interface PreviewExportResult {
  /** Chemin du fichier exporté */
  path: string;
  /** Taille du fichier exporté */
  file_size: number;
  /** Temps d'export en millisecondes */
  export_time: number;
  /** Format utilisé */
  format: PreviewExportFormat;
}

File: src-tauri/src/models/preview.rs
#![allow(dead_code)] // Modèles preview planifiés pour la Phase 2.3

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use uuid::Uuid;

/// Types de previews disponibles (pyramide d'images)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PreviewType {
    /// Thumbnail pour grille (240px bord long, JPEG q75)
    Thumbnail,
    /// Standard pour affichage plein écran (1440px bord long, JPEG q85)
    Standard,
    /// 1:1 pour zoom pixel (résolution native, JPEG q90)
    OneToOne,
}

impl PreviewType {
    pub fn default_size(&self) -> (u32, u32) {
        match self {
            PreviewType::Thumbnail => (240, 240),
            PreviewType::Standard => (1440, 1080),
            PreviewType::OneToOne => (0, 0), // Résolution native
        }
    }

    pub fn jpeg_quality(&self) -> u8 {
        match self {
            PreviewType::Thumbnail => 75,
            PreviewType::Standard => 85,
            PreviewType::OneToOne => 90,
        }
    }

    pub fn max_file_size(&self) -> u64 {
        match self {
            PreviewType::Thumbnail => 50 * 1024,      // 50KB
            PreviewType::Standard => 500 * 1024,      // 500KB
            PreviewType::OneToOne => 2 * 1024 * 1024, // 2MB
        }
    }

    pub fn subdir_name(&self) -> &'static str {
        match self {
            PreviewType::Thumbnail => "thumbnails",
            PreviewType::Standard => "standard",
            PreviewType::OneToOne => "native",
        }
    }
}

/// Configuration pour la génération de previews (pyramide d'images)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewConfig {
    /// Répertoire du catalogue (contient Previews.lrdata/)
    pub catalog_dir: PathBuf,
    /// Nombre de threads pour traitement parallèle
    pub parallel_threads: usize,
    /// Timeout pour génération (secondes)
    pub generation_timeout: u64,
    /// Utiliser libvips si disponible, sinon image crate
    pub use_libvips: bool,
}

impl Default for PreviewConfig {
    fn default() -> Self {
        let mut catalog_dir = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
        catalog_dir.push("Pictures");
        catalog_dir.push("LuminaFast");

        Self {
            catalog_dir,
            parallel_threads: num_cpus::get(),
            generation_timeout: 30,
            use_libvips: false, // Pour l'instant, utiliser image crate
        }
    }
}

impl PreviewConfig {
    /// Récupère le répertoire des previews
    pub fn previews_dir(&self) -> PathBuf {
        self.catalog_dir.join("Previews.lrdata")
    }

    /// Récupère le répertoire pour un type de preview
    pub fn preview_type_dir(&self, preview_type: PreviewType) -> PathBuf {
        self.previews_dir().join(preview_type.subdir_name())
    }
}

/// Résultat de la génération d'une preview
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewResult {
    /// Chemin vers le fichier preview généré
    pub path: PathBuf,
    /// Type de preview
    pub preview_type: PreviewType,
    /// Dimensions réelles (largeur, hauteur)
    pub size: (u32, u32),
    /// Taille du fichier en octets
    pub file_size: u64,
    /// Temps de génération
    pub generation_time: std::time::Duration,
    /// Hash BLAKE3 du fichier original
    pub source_hash: String,
    /// Timestamp de génération
    pub generated_at: DateTime<Utc>,
}

/// Métadonnées d'une preview dans la base de données previews.db
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewRecord {
    /// ID unique de la preview
    pub id: i64,
    /// Hash BLAKE3 du fichier source
    pub source_hash: String,
    /// Type de preview
    pub preview_type: PreviewType,
    /// Chemin relatif depuis Previews.lrdata/
    pub relative_path: String,
    /// Dimensions réelles (largeur, hauteur)
    pub width: u32,
    pub height: u32,
    /// Taille du fichier en octets
    pub file_size: u64,
    /// Qualité JPEG utilisée
    pub jpeg_quality: u8,
    /// Date de génération
    pub generated_at: DateTime<Utc>,
    /// Date de dernier accès
    pub last_accessed: DateTime<Utc>,
    /// Nombre d'accès
    pub access_count: u64,
}

/// Nouvelle preview à insérer dans la base
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NewPreviewRecord {
    pub source_hash: String,
    pub preview_type: PreviewType,
    pub relative_path: String,
    pub width: u32,
    pub height: u32,
    pub file_size: u64,
    pub jpeg_quality: u8,
}

/// Informations sur le cache de previews
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewCacheInfo {
    /// Nombre total de previews en cache
    pub total_previews: usize,
    /// Taille totale du cache (octets)
    pub total_size: u64,
    /// Nombre de thumbnails
    pub thumbnail_count: usize,
    /// Nombre de previews
    pub preview_count: usize,
    /// Dernier cleanup
    pub last_cleanup: Option<DateTime<Utc>>,
}

/// Statistiques de génération batch
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchPreviewStats {
    /// ID unique du batch
    pub batch_id: Uuid,
    /// Nombre total de fichiers traités
    pub total_files: usize,
    /// Nombre de succès
    pub successful_count: usize,
    /// Nombre d'échecs
    pub failed_count: usize,
    /// Nombre de fichiers skip (déjà en cache)
    pub skipped_count: usize,
    /// Temps total de traitement
    pub total_duration: std::time::Duration,
    /// Temps moyen par fichier
    pub avg_time_per_file: std::time::Duration,
    /// Timestamp de début
    pub started_at: DateTime<Utc>,
    /// Timestamp de fin
    pub completed_at: Option<DateTime<Utc>>,
}

/// Erreurs spécifiques au service preview
#[derive(Debug, Clone, Serialize, Deserialize, thiserror::Error)]
pub enum PreviewError {
    #[error("Fichier RAW non supporté: {format}")]
    UnsupportedFormat { format: String },

    #[error("Fichier corrompu ou illisible: {path}")]
    CorruptedFile { path: String },

    #[error("Erreur de traitement RAW: {message}")]
    ProcessingError { message: String },

    #[error("Erreur d'écriture preview: {path}")]
    WriteError { path: String },

    #[error("Erreur cache: {message}")]
    CacheError { message: String },

    #[error("Timeout génération: {timeout}s")]
    GenerationTimeout { timeout: u64 },

    /*     #[error("Mémoire insuffisante pour traitement")]
    OutOfMemory, */
    #[error("Erreur I/O: {message}")]
    IoError { message: String },
}

/// État d'une preview dans le cache
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum PreviewStatus {
    /// Preview générée et valide
    Ready,
    /// En cours de génération
    Generating,
    /// Erreur de génération
    Failed,
    /// Obsolète (fichier source modifié)
    Stale,
}

/// Métadonnées d'une preview en cache
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewMetadata {
    /// Hash BLAKE3 du fichier source
    pub source_hash: String,
    /// Type de preview
    pub preview_type: PreviewType,
    /// Chemin du fichier preview
    pub path: PathBuf,
    /// État actuel
    pub status: PreviewStatus,
    /// Taille du fichier preview
    pub file_size: u64,
    /// Dimensions
    pub size: (u32, u32),
    /// Date de génération
    pub generated_at: DateTime<Utc>,
    /// Dernier accès
    pub last_accessed: DateTime<Utc>,
    /// Nombre d'accès
    pub access_count: u64,
}

/// Événement de progression de génération
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewProgressEvent {
    /// ID unique de l'événement
    pub id: Uuid,
    /// Type de preview
    pub preview_type: PreviewType,
    /// Chemin du fichier source
    pub source_path: PathBuf,
    /// Progression (0.0 à 1.0)
    pub progress: f64,
    /// Message de statut
    pub message: String,
    /// Timestamp
    pub timestamp: DateTime<Utc>,
    /// Statistiques complètes (pour batch_completed)
    pub stats: Option<BatchPreviewStats>,
}

/// Configuration du cache cleanup
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheCleanupConfig {
    /// Taille maximale du cache (octets)
    pub max_cache_size: u64,
    /// Âge maximal des previews (jours)
    pub max_age_days: u32,
    /// Nombre maximal de previews par type
    pub max_previews_per_type: usize,
    /// Fréquence de cleanup (heures)
    pub cleanup_interval_hours: u32,
}

impl Default for CacheCleanupConfig {
    fn default() -> Self {
        Self {
            max_cache_size: 2 * 1024 * 1024 * 1024, // 2GB
            max_age_days: 30,
            max_previews_per_type: 10000,
            cleanup_interval_hours: 24,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::Duration;

    #[test]
    fn test_preview_type_defaults() {
        assert_eq!(PreviewType::Thumbnail.default_size(), (240, 240));
        assert_eq!(PreviewType::Standard.default_size(), (1440, 1080));
        assert_eq!(PreviewType::OneToOne.default_size(), (0, 0)); // Résolution native

        assert_eq!(PreviewType::Thumbnail.jpeg_quality(), 75);
        assert_eq!(PreviewType::Standard.jpeg_quality(), 85);
        assert_eq!(PreviewType::OneToOne.jpeg_quality(), 90);

        assert_eq!(PreviewType::Thumbnail.max_file_size(), 50 * 1024);
        assert_eq!(PreviewType::Standard.max_file_size(), 500 * 1024);
        assert_eq!(PreviewType::OneToOne.max_file_size(), 2 * 1024 * 1024);

        assert_eq!(PreviewType::Thumbnail.subdir_name(), "thumbnails");
        assert_eq!(PreviewType::Standard.subdir_name(), "standard");
        assert_eq!(PreviewType::OneToOne.subdir_name(), "native");
    }

    #[test]
    fn test_preview_config_default() {
        let config = PreviewConfig::default();
        assert_eq!(config.generation_timeout, 30);
        assert!(!config.use_libvips);
        assert!(config.parallel_threads > 0);
        assert!(config.catalog_dir.ends_with("Pictures/LuminaFast"));
        assert!(config.previews_dir().ends_with("Previews.lrdata"));
    }

    #[test]
    fn test_preview_result_serialization() {
        let result = PreviewResult {
            path: PathBuf::from("/test/preview.jpg"),
            preview_type: PreviewType::Thumbnail,
            size: (240, 180),
            file_size: 1024,
            generation_time: Duration::from_millis(100),
            source_hash: "b3a1c2d3".to_string(),
            generated_at: Utc::now(),
        };

        let serialized = serde_json::to_string(&result).unwrap();
        let deserialized: PreviewResult = serde_json::from_str(&serialized).unwrap();

        assert_eq!(result.preview_type, deserialized.preview_type);
        assert_eq!(result.size, deserialized.size);
        assert_eq!(result.file_size, deserialized.file_size);
    }

    #[test]
    fn test_preview_error_display() {
        let error = PreviewError::UnsupportedFormat {
            format: "cr3".to_string(),
        };
        assert!(error.to_string().contains("cr3"));

        let error = PreviewError::GenerationTimeout { timeout: 30u64 };
        assert!(error.to_string().contains("30s"));
    }

    #[test]
    fn test_cache_cleanup_config_default() {
        let config = CacheCleanupConfig::default();
        assert_eq!(config.max_cache_size, 2 * 1024 * 1024 * 1024);
        assert_eq!(config.max_age_days, 30);
        assert_eq!(config.max_previews_per_type, 10000);
        assert_eq!(config.cleanup_interval_hours, 24);
    }
}
